File: mobile/apps/parent/app/src/main/kotlin/com/schoolerp/parent/PullToRefresh.kt
package com.schoolerp.parent

import android.content.Context
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import android.view.animation.DecelerateInterpolator
import android.widget.FrameLayout
import android.widget.ProgressBar

/* PULL DOWN TO RELOAD, BECAUSE A PAGE THAT CANNOT BE PULLED READS AS A PAGE.

   The gesture is close to the only thing every Android user tests without
   meaning to. A screen that does not answer it is a screen they conclude is
   stale, and the usual next move is to close the app and open it again, which
   in a shell like this one costs a full cold start and a re round trip for
   everything. Answering the pull is cheaper than the workaround it prevents.

   The obvious implementation is androidx SwipeRefreshLayout. It is not used
   here for the reason given in app/build.gradle.kts: this app has no
   dependencies, and a single gesture is not worth starting a version train
   the tracker app would then have to be kept in step with. This is that one
   gesture and nothing else, at roughly a tenth of the size.

   THE HARD PART IS NOT THE ANIMATION, IT IS KNOWING WHEN TO KEEP OUT OF THE
   WAY. The thing being wrapped is a whole web application, and the failure
   mode of a naive version is that it steals downward drags from the page: a
   parent scrolling a fee ledger back up to the top finds the app reloading
   under them and losing their place, which is much worse than having no
   gesture at all. So the rules here are deliberately mean, and every one of
   them exists to refuse rather than to trigger:

     - It only ever considers a gesture that began with the content already at
       the very top. canScrollUp is asked once, at ACTION_DOWN, and not again,
       because asking mid gesture is how a fling that lands at zero turns into
       a refresh nobody asked for.
     - It only considers a drag that is mostly vertical. A horizontal swipe on
       a carousel or a table that scrolls sideways is not a pull, and comparing
       the two deltas is the cheapest way to say so.
     - It requires the finger to travel a full touch slop downward before it
       takes the gesture, so a tap that wobbles is still a tap.
     - It gives up the instant the finger goes back above where it started.
       A parent who begins a pull and changes their mind gets their scroll
       back rather than a rubber band that will not let go.

   What it deliberately cannot do is see inside the page. If the site puts its
   own scrolling container inside the viewport, the WebView itself is at scroll
   zero the whole time and this will offer to refresh while the parent is
   scrolling that inner list. There is no native side answer to that; the fix
   if it ever bites is for the page to let the document scroll rather than a
   div. It is written down here so the next person does not have to rediscover
   it from a bug report.
*/
class PullToRefresh(context: Context) : FrameLayout(context) {

    /* Set by the activity. Returns true while the wrapped content is scrolled
       away from its top, in which case this layout never takes a gesture. */
    var canScrollUp: () -> Boolean = { false }

    /* Called once per accepted pull. The activity reloads; it must call
       stopRefreshing when the load finishes, or the spinner sits there
       forever and the app looks hung. */
    var onRefresh: () -> Unit = {}

    /* Turned off while an error panel is showing. The panel has its own
       explicit retry button, and a pull that silently did the same thing
       behind it left two competing ways to do one act, one of them invisible.

       Named pullEnabled rather than the obvious "enabled": View already has a
       setEnabled, and a Kotlin property of that name silently becomes an
       accidental override of it, which would have meant switching the gesture
       off also switching off touch handling for the WebView underneath. The
       compiler caught it; the name is kept distinct so nobody reintroduces
       it. */
    var pullEnabled = true

    private val spinner = ProgressBar(context).apply {
        isIndeterminate = true
        visibility = View.INVISIBLE
    }
    private val slop = ViewConfiguration.get(context).scaledTouchSlop
    private val density = context.resources.displayMetrics.density
    private val trigger = 72f * density
    private val ceiling = 132f * density
    private val resting = -40f * density

    private var startX = 0f
    private var startY = 0f
    private var eligible = false
    private var dragging = false
    private var refreshing = false

    init {
        val size = (28 * density).toInt()
        addView(
            spinner,
            LayoutParams(size, size, android.view.Gravity.TOP or android.view.Gravity.CENTER_HORIZONTAL)
                .apply { topMargin = (16 * density).toInt() },
        )
        spinner.translationY = resting
    }

    /* Intercept rather than handle. While this returns false every touch goes
       straight through to the WebView untouched, which is the state it is in
       for all but a deliberate pull, and is why ordinary scrolling is not
       affected at all. */
    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
        if (!pullEnabled || refreshing) return false
        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                startX = event.x
                startY = event.y
                dragging = false
                // The one and only reading of scroll position, taken before a
                // finger has moved anything.
                eligible = !canScrollUp()
            }
            MotionEvent.ACTION_POINTER_DOWN -> {
                // A second finger means a pinch zoom on a document, never a
                // pull. Hand the whole gesture back.
                eligible = false
                dragging = false
            }
            MotionEvent.ACTION_MOVE -> {
                if (!eligible) return false
                val dy = event.y - startY
                val dx = event.x - startX
                if (dy > slop && dy > Math.abs(dx)) {
                    dragging = true
                    return true
                }
                // Any upward or sideways travel past slop settles the question
                // for the rest of this gesture: it was not a pull.
                if (dy < -slop || Math.abs(dx) > slop) eligible = false
            }
        }
        return false
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!dragging) return false
        when (event.actionMasked) {
            MotionEvent.ACTION_MOVE -> {
                val dy = event.y - startY - slop
                if (dy <= 0f) {
                    // Back above the start. Abandon rather than clamp at zero,
                    // so the parent gets their scrolling back mid gesture.
                    release(false)
                    return false
                }
                // Damped, so the spinner slows as it is pulled and the travel
                // has a floor. An undamped follow makes a short flick look
                // like a committed pull.
                spinner.translationY = resting + Math.min(dy * 0.5f, ceiling)
                spinner.visibility = View.VISIBLE
                spinner.alpha = Math.min(1f, dy / trigger)
                return true
            }
            MotionEvent.ACTION_UP -> {
                release(spinner.translationY - resting >= trigger * 0.5f)
                return true
            }
            MotionEvent.ACTION_CANCEL -> {
                release(false)
                return true
            }
        }
        return true
    }

    private fun release(commit: Boolean) {
        dragging = false
        eligible = false
        if (!commit) {
            retract()
            return
        }
        refreshing = true
        spinner.animate().translationY(24f * density).alpha(1f).setDuration(150).start()
        onRefresh()
    }

    /* Called by the activity when the load it started has finished, failed, or
       been replaced. Idempotent on purpose: a page that fires both an error
       and a finish would otherwise animate twice. */
    fun stopRefreshing() {
        if (!refreshing) return
        refreshing = false
        retract()
    }

    private fun retract() {
        spinner.animate()
            .translationY(resting)
            .alpha(0f)
            .setInterpolator(DecelerateInterpolator())
            .setDuration(200)
            .withEndAction { spinner.visibility = View.INVISIBLE }
            .start()
    }
}

File: mobile/apps/parent-ios/ParentApp/Shell/PullToRefresh.swift
import UIKit

/* PULL DOWN TO RELOAD, BECAUSE A PAGE THAT CANNOT BE PULLED READS AS A PAGE.

   The obvious implementation is UIRefreshControl on the web view's scroll
   view. It is not used, for the reason the Android PullToRefresh records
   about SwipeRefreshLayout: the thing being wrapped is a whole web
   application that scrolls an element INSIDE the document, so the scroll view
   sits at zero on nearly every screen however far down the parent actually
   is, and a refresh control on it fires fifteen screens into the setup
   wizard. On a half-filled form that loses the form.

   The page tells the shell where its own scroller is (setAtTop, through the
   bridge) and this gesture believes the page. The rules are deliberately
   mean, and every one exists to refuse rather than to trigger:

     - it only considers a gesture that began with the page reporting itself
       at the top, read once when the pan begins and not again;
     - it only considers a drag that is mostly vertical;
     - it requires a full slop of downward travel before it takes the gesture;
     - it gives up the instant the finger goes back above where it started.

   The recogniser runs alongside the web view's own, never instead of it, so
   the page still scrolls and bounces underneath and ordinary use is not
   affected at all. */
final class PullToRefresh: NSObject, UIGestureRecognizerDelegate {

    /* Set by the shell. Returns true while the page reports its scroller is
       away from the top, in which case no gesture is ever taken. */
    var canScrollUp: () -> Bool = { true }

    /* Called once per accepted pull. The shell reloads; it must call
       stopRefreshing when the load finishes, or the spinner sits there for
       ever and the app looks hung. */
    var onRefresh: () -> Void = {}

    /* The spinner's travel and opacity, for the SwiftUI layer to draw. */
    var onChange: (_ travel: CGFloat, _ alpha: CGFloat, _ visible: Bool) -> Void = { _, _, _ in }

    /* Turned off while an error panel is showing. The panel has its own retry
       button, and a pull that silently did the same thing behind it is two
       competing ways to do one act, one of them invisible. */
    var pullEnabled = true

    private(set) var refreshing = false

    let recognizer = UIPanGestureRecognizer()

    private let slop: CGFloat = 10
    private let trigger: CGFloat = 72
    private let ceiling: CGFloat = 132

    private var eligible = false
    private var dragging = false
    private var travel: CGFloat = 0

    override init() {
        super.init()
        recognizer.addTarget(self, action: #selector(pan(_:)))
        recognizer.maximumNumberOfTouches = 1
        recognizer.cancelsTouchesInView = false
        recognizer.delegate = self
    }

    func gestureRecognizer(
        _ gestureRecognizer: UIGestureRecognizer,
        shouldRecognizeSimultaneouslyWith other: UIGestureRecognizer
    ) -> Bool {
        true
    }

    @objc private func pan(_ g: UIPanGestureRecognizer) {
        switch g.state {
        case .began:
            dragging = false
            travel = 0
            // The one and only reading of the page's position.
            eligible = pullEnabled && !refreshing && !canScrollUp()
            // A pan begins only after the finger has already moved, so the
            // first translation is worth judging rather than discarding.
            moved(g)
        case .changed:
            moved(g)
        case .ended:
            if dragging { release(travel >= trigger * 0.5) }
            eligible = false
        case .cancelled, .failed:
            release(false)
        default:
            break
        }
    }

    private func moved(_ g: UIPanGestureRecognizer) {
        guard eligible else { return }
        let t = g.translation(in: g.view)
        if !dragging {
            if t.y > slop && t.y > abs(t.x) {
                dragging = true
            } else if t.y < -slop || abs(t.x) > slop {
                // Any upward or sideways travel past slop settles the question
                // for the rest of this gesture: it was not a pull.
                eligible = false
                return
            } else {
                return
            }
        }
        let dy = t.y - slop
        if dy <= 0 {
            // Back above the start. Abandon rather than clamp at zero, so the
            // parent gets their scrolling back mid gesture.
            release(false)
            return
        }
        // Damped, so the spinner slows as it is pulled and the travel has a
        // floor. An undamped follow makes a short flick look like a pull.
        travel = min(dy * 0.5, ceiling)
        onChange(travel, min(1, dy / trigger), true)
    }

    private func release(_ commit: Bool) {
        dragging = false
        eligible = false
        if !commit {
            onChange(0, 0, false)
            return
        }
        refreshing = true
        onChange(64, 1, true)
        onRefresh()
    }

    /* Called by the shell when the load it started has finished, failed, or
       been replaced. Idempotent on purpose: a page that fires both an error
       and a finish would otherwise animate twice. */
    func stopRefreshing() {
        if !refreshing { return }
        refreshing = false
        onChange(0, 0, false)
    }
}
